### PR TITLE
Wait for adblock components to be fully initialized before loading engines

### PR DESCRIPTION
**Pull request by antonok-edm**
Resolves https://github.com/brave/brave-browser/issues/35922

While tracing https://github.com/brave/brave-core/pull/22049, I noticed that `ReadDATFileData` is called over 2000 times when all adblock lists are enabled. This change cuts that down to 53 (i.e. just once per component).

**NOTE**: this change increases the amount of time necessary for adblock to be available after first launch, limiting it by the time taken to download the **slowest** adblock component. I was able to see YouTube ads on a fresh profile by immediately pasting `youtube.com` into the URL bar after making this change. It's not obvious how likely that is to be an issue across various network conditions and launch patterns. That said, this change _does_ noticeably improve startup performance, and I'm running out of alternative options.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_shields/core/browser/ad_block_filters_provider.cc [hunk 1]
```diff
@@ -45,6 +45,10 @@ void AdBlockFiltersProvider::NotifyObservers(bool is_for_default_engine) {
   }
 }
 
+bool AdBlockFiltersProvider::IsInitialized() {
+  return true;
```
Comment (bridiver) [inline components/brave_shields/core/browser/ad_block_filters_provider.cc:49]: why is the default true instead of false?

Comment (bridiver) [reply components/brave_shields/core/browser/ad_block_filters_provider.cc:49]: I guess this makes sense https://github.com/brave/brave-core/pull/22082/files#diff-4d045535d239d0bdab24eefe097d8f98e4d0d4d12c37714a0e3f34b387e94420R49

**Comment by atuchin-m** [comment]
@antonok-edm 
Do we still have the issue with CPU usage and excessive `ReadDATFileData` calls during the components update scenario?
If update is pending is `OnComponentReady()` called one before the update done (and then after)?

Taking into the account list updating frequency the usual startup scenario is:
1) parse/load the current (old) components from the disk.
2) fetch the updates
3) apply the updates

I'm worry about a bunch of `ReadDATFileData` calls  on stage 3.  All `IsInitialized()` are `true` that time, the PR couldn't prevent any excessive work on that stage.

**Review by atuchin-m — APPROVED** [review]

**Comment by kjozwiak** [comment]
**`Quick Note:`** For each of the cases mentioned below, the following was done in terms of STR/Cases:

* ensured that `https://www.buzzfeed.com/ca` was already copied/ready to paste into URL bar
* ensured that all profiles have been removed
* launched Brave and then quickly pasted `https://www.buzzfeed.com/ca` and resolved (as quickly as possible)

If an issue was found/ads initially being displayed, compared against `1.63.169 Chromium: 122.0.6261.111` and ensured that it wasn't any worse with the above PR. Basically ensuring it was also happening via the current `release` build.

### `Win 10 x64` verifications using a VM (slowing down VM & Network)

#### `Test Case #1` - Slowing down VM via `RAM`  & `CPU` - **`No ads appearing on initial launch`**

* set the `RAM` of the current VM to `4GB`
* assigned `2` CPU's to the VM from host (ensured that `Dynamic Memory` is also disabled)

**`Example`** | **`Example`**
----------------|----------------
image | image

https://github.com/brave/brave-core/assets/2602313/cc8d657d-bec3-4fdf-973f-e905ec381b1e

#### `Test Case #2` - Slowing down VM via `CPU`, `RAM` & `Network` (`200ms` delay) - **`No ads appearing on initial launch`**

* set the `RAM` of the current VM to `4GB`
* assigned `2` CPU's to the VM from host (ensured that `Dynamic Memory` is also disabled)
* using https://jagt.github.io/clumsy/index.html to slow down the VM network (using `200ms` delay)

https://github.com/brave/brave-core/assets/2602313/94b6dda7-37a1-42d3-afa5-1b7809845d43

#### `Test Case #3` - Slowing down VM via `CPU`, `RAM` & `Network` (`300ms` delay) - **`ads appearing on initial launch`**

* set the `RAM` of the current VM to `4GB`
* assigned `2` CPU's to the VM from host (ensured that `Dynamic Memory` is also disabled)
* using https://jagt.github.io/clumsy/index.html to slow down the VM network (using `300ms` delay)

With the above case, you'll notice that some of the ads do end up appearing when slowing down the network via `300ms` but it also happens with `1.63.169 Chromium: 122.0.6261.111` so it's not any better/nor any worse.

https://github.com/brave/brave-core/assets/2602313/b0bdb4e2-2cef-40b7-bc15-be30da862d9f

**Pull request merged**